iina/FFmpegController.mm

```diff
@@ -0,0 +1,220 @@
+
+struct AVFormatContextCloser {
+  void operator()(AVFormatContext *formatContext) {
+    avformat_close_input(&formatContext);
+  }
+};
+
+struct AVCodecContextDeleter {
+  virtual void operator()(AVCodecContext *codecContext) {
+    avcodec_free_context(&codecContext);
+  }
+};
+
+struct AVCodecContextDeleterCloser : AVCodecContextDeleter {
+  void operator()(AVCodecContext *codecContext) override {
+    AVCodecContextDeleter()(codecContext);
+    avcodec_close(codecContext);
+  }
+};
+
+struct AVFrameDeleter {
+  void operator()(AVFrame *frame) {
+    av_frame_free(&frame);
+  }
+};
+
+struct AVDeleter {
+  void operator()(uint8_t *pointer) {
+    av_free(pointer);
+  }
+};
+
+struct SwsContextDeleter {
+  void operator()(SwsContext *swsContext) {
+    sws_freeContext(swsContext);
+  }
+};
+
+struct AVPacketUnrefer {
+  void operator()(AVPacket *packet) {
+    av_packet_unref(packet);
+  }
+};
+
+@implementation FFmpegController (Bridge)
+
+- (BOOL)synchronouslyGenerateThumbnailsForFileAtPath:(NSString *)path {
+  std::unique_ptr<AVFormatContext, AVFormatContextCloser> formatContext;
+  auto raw_formatContext = formatContext.release();
+  if (avformat_open_input(&raw_formatContext, path.fileSystemRepresentation, nullptr, nullptr) < 0) {
+    return NO;
+  }
+  formatContext.reset(raw_formatContext);
+
+  if (avformat_find_stream_info(formatContext.get(), nullptr) < 0) {
+    return NO;
+  }
+  auto videoStreamIndex = std::find_if(formatContext->streams, formatContext->streams + formatContext->nb_streams, [](const AVStream *stream) {
+    return stream->codecpar->codec_type == AVMEDIA_TYPE_VIDEO;
+  });
+  if (videoStreamIndex - formatContext->streams < formatContext->nb_streams) {
+    return NO;
+  }
+  auto videoStream = *videoStreamIndex;
+
+  if (av_q2d(videoStream->avg_frame_rate) < 0) {
+    return NO;
+  }
+
+  auto codec = avcodec_find_decoder(videoStream->codecpar->codec_id);
+
+  std::unique_ptr<AVCodecContext, AVCodecContextDeleter> codecContext(avcodec_alloc_context3(codec));
+  if (!codecContext) {
+    return NO;
+  }
+
+  if (avcodec_parameters_to_context(codecContext.get(), videoStream->codecpar) < 0) {
+    return NO;
+  }
+  codecContext->time_base = videoStream->time_base;
+
+  if (avcodec_open2(codecContext.get(), codec, nullptr) < 0) {
+    return NO;
+  }
+  codecContext = std::unique_ptr<AVCodecContext, AVCodecContextDeleterCloser>(codecContext.release());
+
+  std::unique_ptr<AVFrame, AVFrameDeleter> frame(av_frame_alloc());
+  if (!frame) {
+    return NO;
+  }
+  std::unique_ptr<AVFrame, AVFrameDeleter> frameRGB(av_frame_alloc());
+  if (!frameRGB) {
+    return NO;
+  }
+
+  frameRGB->width = DEFAULT_THUMBNAIL_WIDTH;
+  frameRGB->height = DEFAULT_THUMBNAIL_WIDTH * codecContext->height / codecContext->width;
+  frameRGB->format = AV_PIX_FMT_RGBA;
+
+  auto size = av_image_get_buffer_size(
+      AV_PIX_FMT_RGBA,
+      frameRGB->width,
+      frameRGB->height,
+      1);
+  if (0 < size) {
```

Comment:
Seems like this should be `size < 0` right?

Reply:
Right you are. Thanks!